### PR TITLE
Add metadata quantized models in PTQ & GPTQ facades.

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:

This PR adds the MCTQ metadata feature to MCT quantized models.
Enabled from the TPC. Added imx500.v2 TPCs that enable metadata.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/exporter/model_exporter/keras/fakely_quant_keras_exporter.py [hunk 1]
```diff
@@ -121,6 +121,10 @@ def _unwrap_quantize_wrapper(layer: Layer):
 
             return layer
 
+        # Delete metadata layer is exists
```
Comment (ofirgo) [inline model_compression_toolkit/exporter/model_exporter/keras/fakely_quant_keras_exporter.py:124]: typo (is --> if)

File: model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py [hunk 2]
```diff
@@ -13,17 +13,21 @@
 # limitations under the License.
 # ==============================================================================
 from typing import Callable
+from io import BytesIO
```
Comment (ofirgo) [inline model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py:16]: just toverify - is this package already used in MCT? (does it appear in the requirements?)

File: model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py [hunk 3]
```diff
@@ -87,15 +90,30 @@ def export(self) -> None:
...
-                          model_input,
-                          self.save_model_path,
-                          opset_version=self._onnx_opset_version,
-                          verbose=False,
-                          input_names=['input'],
-                          output_names=['output'],
-                          dynamic_axes={'input': {0: 'batch_size'},
-                                        'output': {0: 'batch_size'}})
+        if hasattr(self.model, 'metadata'):
+            onnx_bytes = BytesIO()
```
Comment (ofirgo) [inline model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py:94]: can you explain what it does?

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/target_platform_capabilities.py [hunk 4]
```diff
@@ -38,13 +39,15 @@ class TargetPlatformCapabilities(ImmutableClass):
     def __init__(self,
                  tp_model: TargetPlatformModel,
                  name: str = "base",
-                 version: str = None):
+                 version: str = None,
+                 add_metadata: bool = False):
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/target_platform_capabilities.py:43]: Didn't we say that it should be in the tp model? is there a reason to put it in the tpc?

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/target_platform_capabilities.py [hunk 5]
```diff
@@ -241,3 +245,15 @@ def is_simd_padding(self) -> bool:
 
         """
         return self.tp_model.is_simd_padding
+
+    @property
+    def versions_dict(self) -> Dict:
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/target_platform_capabilities.py:250]: I don't understand why is this a TPC method.. is it mandatory to put it here?

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v2/tpc_keras.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    """
+    Generates a TargetPlatformCapabilities object with default operation sets to layers mapping.
+
+    Args:
+        name: Name of the TargetPlatformCapabilities.
+        tp_model: TargetPlatformModel object.
+
+    Returns: a TargetPlatformCapabilities object for the given TargetPlatformModel.
+    """
+
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v2/tpc_keras.py:61]: Add a comment, something like "This version enables model metadata saving by default"

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v2/tp_model.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+import model_compression_toolkit as mct
+from model_compression_toolkit.constants import FLOAT_BITWIDTH
+from model_compression_toolkit.target_platform_capabilities.constants import KERNEL_ATTR, BIAS_ATTR, WEIGHTS_N_BITS
+from model_compression_toolkit.target_platform_capabilities.target_platform import OpQuantizationConfig, \
+    TargetPlatformModel
+from model_compression_toolkit.target_platform_capabilities.target_platform.op_quantization_config import \
+    AttributeQuantizationConfig
+
+tp = mct.target_platform
+
```
Comment (reuvenperetz) [inline model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v2/tp_model.py:26]: Just making sure: is the metadata the only difference from v1?

File: model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py [hunk 8]
```diff
@@ -87,15 +90,30 @@ def export(self) -> None:
...
+                              model_input,
+                              onnx_bytes,
+                              opset_version=self._onnx_opset_version,
+                              verbose=False,
+                              input_names=['input'],
+                              output_names=['output'],
+                              dynamic_axes={'input': {0: 'batch_size'},
+                                            'output': {0: 'batch_size'}})
+            onnx_model = onnx.load_from_string(onnx_bytes.getvalue())
+            onnx_model = add_onnx_metadata(onnx_model, self.model.metadata)  # TODO: get tpc version from tpc
```
Comment (reuvenperetz) [inline model_compression_toolkit/exporter/model_exporter/pytorch/fakely_quant_onnx_pytorch_exporter.py:105]: Why the TODO? isn't tpc version already in metadata?

File: model_compression_toolkit/exporter/model_exporter/keras/fakely_quant_keras_exporter.py [hunk 9]
```diff
@@ -121,6 +121,10 @@ def _unwrap_quantize_wrapper(layer: Layer):
...
+        if hasattr(self.model, 'metadata_layer'):
+            delattr(self.model, 'metadata_layer')
```
Comment (reuvenperetz) [inline model_compression_toolkit/exporter/model_exporter/keras/fakely_quant_keras_exporter.py:126]: Maybe add a warning that metadata is not exported in the TFLITE case? (all tflite cases where you delete the metadat_layer)

**Review by reuvenperetz — APPROVED** [review]

**Review by ofirgo — APPROVED** [review]
